### PR TITLE
server/coordinator: fix add operator

### PR DESCRIPTION
**Pull request by nolouch**
Currently, when we add a priority kind operator, the old operator limiter no remove. the limiter will always increase when change from other kind operators to priority kind operator.the `AllowScheduler` will block the scheduler. 
PTAL  @disksing @andelf @siddontang

### CODE REVIEW

File: server/coordinator.go [hunk 1]
```diff
@@ -211,8 +211,12 @@ func (c *coordinator) addOperator(op Operator) bool {
...
 	if op.GetResourceKind() == priorityKind {
 		o, ok := c.operators[regionID]
-		if ok && o.GetResourceKind() == priorityKind {
-			return false
+		if ok {
+			if o.GetResourceKind() == priorityKind {
+				return false
+			}
+			c.limiter.removeOperator(o)
+			log.Infof("coordinator: add more prioriy operator: %+v remove operator: %+v", op, o)
```
Comment (disksing) [inline server/coordinator.go]: add operator with higher priority

Comment (disksing) [inline server/coordinator.go]: I think put this block inside `if _, ok := c.operators[regionID]; ok` will make it more clear.

```
if old, ok = c.operator[regionId]; ok {
        if op.GetResourceKind() == priorityKind && old.GetResourceKind() != priorityKind {
                c.limiter. removeOperator(old)
                log.Infof("coordinator: add operator with higher priority: %+v remove operator: %+v", op, o)
        } else {
                return false
        }
}

c.limiter.addOperator(op)
...
```

File: server/coordinator.go [hunk 2]
```diff
@@ -211,8 +211,12 @@ func (c *coordinator) addOperator(op Operator) bool {
 	}
 	if op.GetResourceKind() == priorityKind {
 		o, ok := c.operators[regionID]
-		if ok && o.GetResourceKind() == priorityKind {
-			return false
+		if ok {
+			if o.GetResourceKind() == priorityKind {
+				return false
+			}
+			c.limiter.removeOperator(o)
```
Comment (siddontang) [inline server/coordinator.go]: I guess we should also remove limiter at https://github.com/pingcap/pd/blob/c3ca6537b8945ec4f0040067dd6a16d2af0eefc1/server/coordinator.go#L209

**Comment by disksing** [comment]
LGTM.

File: server/coordinator.go [hunk 3]
```diff
@@ -226,24 +226,14 @@ func (c *coordinator) addOperator(op Operator) bool {
...
-	if op.GetResourceKind() == adminKind {
-		c.operators[regionID] = op
-		return true
-	}
-	if op.GetResourceKind() == priorityKind {
-		o, ok := c.operators[regionID]
-		if ok && o.GetResourceKind() == priorityKind {
+	if old, ok := c.operators[regionID]; ok {
+		if op.GetResourceKind() == adminKind ||
+			(op.GetResourceKind() == priorityKind && old.GetResourceKind() != priorityKind) {
```
Comment (siddontang) [inline server/coordinator.go]: do we all cover these two conditions?

File: server/coordinator.go [hunk 4]
```diff
@@ -226,24 +226,14 @@ func (c *coordinator) addOperator(op Operator) bool {
...
+			c.limiter.removeOperator(old)
+			log.Infof("coordinator: add operator with higher priority: %+v remove operator: %+v", op, old)
```
Comment (siddontang) [inline server/coordinator.go]: `add operator %+v with higher priority, remove operator: %+v`

File: server/coordinator.go [hunk 5]
```diff
@@ -226,24 +226,14 @@ func (c *coordinator) addOperator(op Operator) bool {
...
-	// Admin operator bypasses the check.
-	if op.GetResourceKind() == adminKind {
-		c.operators[regionID] = op
-		return true
-	}
-	if op.GetResourceKind() == priorityKind {
-		o, ok := c.operators[regionID]
-		if ok && o.GetResourceKind() == priorityKind {
+	if old, ok := c.operators[regionID]; ok {
+		if op.GetResourceKind() == adminKind ||
```
Comment (zhangjinpeng1987) [inline server/coordinator.go]: I think we should extract a function to do the judgement work.

Comment (siddontang) [inline server/coordinator.go]: any update? @nolouch

File: server/coordinator.go [hunk 6]
```diff
@@ -252,6 +240,16 @@ func (c *coordinator) addOperator(op Operator) bool {
 	return true
 }
 
+func isPriorityOperator(new Operator, old Operator) bool {
```
Comment (disksing) [inline server/coordinator.go]: How about `shouldReplace`? @zhangjinpeng1987

Comment (siddontang) [inline server/coordinator.go]: This function only checks priority. maybe isHighPriorityOperator

Comment (siddontang) [inline server/coordinator.go]: Btw, we should add an individual test for this function.

**Comment by siddontang** [comment]
Rest LGTM

File: server/coordinator.go (hunk 6 above)
Comment (nolouch) [inline server/coordinator.go]: I think the test only do one thing now

File: server/coordinator_test.go [hunk 7]
```diff
@@ -356,18 +356,34 @@ func (s *testScheduleControllerSuite) TestController(c *C) {
 
 	// limit = 2
 	lb.limit = 2
-	op := newTestOperator(1, leaderKind)
 	// count = 0
 	c.Assert(sc.AllowSchedule(), IsTrue)
-	sc.limiter.addOperator(op)
+	op1 := newTestOperator(1, leaderKind)
+	co.addOperator(op1)
```
Comment (siddontang) [inline server/coordinator_test.go]: please check addOperator return value

**Pull request merged**